autoendpoint/src/server/extractors/user.rs

```diff
@@ -5,31 +5,59 @@ use crate::server::ServerState;
 use autopush_common::db::{DynamoDbUser, DynamoStorage};
 use cadence::{Counted, StatsdClient};
 use futures::compat::Future01CompatExt;
+use std::str::FromStr;
 use uuid::Uuid;
 
 /// Valid `DynamoDbUser::router_type` values
-const VALID_ROUTERS: [&str; 5] = ["webpush", "gcm", "fcm", "apns", "adm"];
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum RouterType {
+    WebPush,
+    GCM,
+    FCM,
+    APNS,
+    ADM,
+}
+
+impl FromStr for RouterType {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "webpush" => Ok(RouterType::WebPush),
+            "gcm" => Ok(RouterType::GCM),
+            "fcm" => Ok(RouterType::FCM),
+            "apns" => Ok(RouterType::APNS),
+            "adm" => Ok(RouterType::ADM),
+            _ => Err(()),
+        }
+    }
+}
```

Comment:
This is such a common pattern, I'm surprised there's not a macro for it.

Reply:
There are third-party derives available, but I'm not sure they're worth the extra dependencies and compile time.
https://docs.rs/enum-utils/0.1.1/enum_utils/derive.FromStr.html